pkg/smokescreen/acl/v1/acl.go

```diff
@@ -0,0 +1,230 @@
+
+	err = acl.ValidateDomains(r.DomainGlobs)
+	if err != nil {
+		return err
+	}
+
+	acl.Rules[svc] = r
+	return nil
+}
+
+// Decide takes uses the rule configured for the given service to determine if
+//   1. The host is in the rule's allowed domain
+//   2. The host has been globally denied
+//   3. The host has been globally allowed
+//   4. There is a default rule for the ACL
+func (acl *ACL) Decide(service, host string) (Decision, error) {
+	var d Decision
+
+	rule := acl.Rule(service)
+	if rule == nil {
+		d.Result = Deny
+		d.Reason = "no rule matched"
+		return d, nil
+	}
+
+	d.Project = rule.Project
+	d.Default = rule == acl.DefaultRule
+
+	// if the host matches any of the rule's allowed domains, allow
+	for _, dg := range rule.DomainGlobs {
+		if hostMatchesGlob(host, dg) {
+			d.Result, d.Reason = Allow, "host matched allowed domain in rule"
+			return d, nil
+		}
+	}
+
+	// if the host matches any of the global deny list, deny
+	for _, dg := range acl.GlobalDenyList {
+		if hostMatchesGlob(host, dg) {
+			d.Result, d.Reason = Deny, "host matched rule in global deny list"
+			return d, nil
+		}
+	}
+
+	// if the host matches any of the global allow list, allow
+	for _, dg := range acl.GlobalAllowList {
+		if hostMatchesGlob(host, dg) {
+			d.Result, d.Reason = Allow, "host matched rule in global allow list"
+			return d, nil
+		}
+	}
+
+	var err error
+	switch rule.Policy {
+	case Report:
+		d.Result, d.Reason = AllowAndReport, "rule has allow and report policy"
+	case Enforce:
+		d.Result, d.Reason = Deny, "rule has enforce policy"
+	case Open:
+		d.Result, d.Reason = Allow, "rule has open enforcement policy"
+	default:
+		d.Result, d.Reason = Unknown, "unexpected policy value"
+		err = fmt.Errorf("unexpected policy value for (%s -> %s): %d", service, host, rule.Policy)
+	}
+
+	if d.Default {
+		d.Reason = "default rule policy used"
+	}
+
+	return d, err
+}
+
+// DisablePolicies takes a slice of actions (open, report, enforce), maps them
+// to their corresponding EnforcementPolicy, and adds them to the global
+// disabledPolicy slice.
+func (acl *ACL) DisablePolicies(actions []string) error {
+	for _, a := range actions {
+		p, err := PolicyFromAction(a)
+		if err != nil {
+			return err
+		}
+		acl.DisabledPolicies = append(acl.DisabledPolicies, p)
+	}
+	return nil
+}
+
+// Validate checks that the ACL that every rule has a conformant domain glob
+// and is not utilizing a disabled enforcement policy.
+func (acl *ACL) Validate() error {
+	for svc, r := range acl.Rules {
+		err := acl.ValidateDomains(r.DomainGlobs)
+		if err != nil {
+			return err
+		}
+		err = acl.PolicyDisabled(svc, r.Policy)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ValidateDomains takes a slice of domains and verifies they conform to
+// smokescreen's domain glob policy.
+//
+// Domains can only contain a single wildcard prefix
+// Domains cannot be represented as a sole wildcard
+func (acl *ACL) ValidateDomains(domains []string) error {
```

Comment:
Since we're iterating through a list of domains here, should we include the domains in the error messages?

Reply:
👍